### ISSUE TITLE
Error: decorator is not a function

### ISSUE THREAD

**Issue by sondreb** [weichx/cerialize#22]
When I try to use the @deserializeAs(Type) it fails with the error:

`decorator is not a function`

Code:

```
export class User {
    @deserialize id: string;
    @deserialize firstName: string;
    @deserialize lastName: string;
    @deserializeAs(Address) address: Address;

    displayName(): string {
        return this.firstName + ' ' + this.lastName;
    }
}

export class Address {
    @deserialize city: string;
    @deserialize country: string;
}
```

Any ideas?

**Comment by weichx** [comment]
Everything looks fine from the code, it might be something else in the environment. What browser and version of typescript are you using? Did you enable the `useExperimentalDecorators` flag in tsConfig?

**Comment by sondreb** [comment]
Yes, it's enabled. Here is the full tsconfig.json:

```
{
  "compilerOptions": {
    "target": "es5",
    "module": "system",
    "moduleResolution": "node",
    "sourceMap": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "removeComments": false,
    "noImplicitAny": false,
    "jsx": "react"
  },
  "exclude": [
    "node_modules",
    "platforms",
    "www"
  ]
}
```

The es6-shim is loaded, file es6-shim.min.js.

Code works fine if I remove the @deserializeAs().

Compiled with TypeScript v1.8.10 and built using gulp-typescript v2.13.0

I have used similar annotations (with parameters) in another TypeScript library I'm building, and there they work just fine, but that is Node.js and ES6 output, and not Angular 2.

**Comment by weichx** [comment]
Hmm, where is the source of the decorator is not a function error? Is that coming from within Cerialize?

**Comment by sondreb** [comment]
The error is thrown from Angular boot upon loading of the app, so long time before I run any serialize/deserialize operations in Cerialize. So probably when Angular parses/runs the attributes defined on the class.

Upon loading I'm also getting this error, might be related as it has to do with Map?

```
serialize.js:10 Uncaught ReferenceError: exports is not defined
```

Here is the code in question:

![image](https://cloud.githubusercontent.com/assets/309938/15227281/b176a722-1886-11e6-8cd5-4287c4053eb4.png)


We are not on the latest builds of Angular and some dependent libs, here is the list:

```
    "adal-angular": "1.0.9",
    "angular2": "2.0.0-beta.3",
    "cerialize": "0.1.5",
    "es6-promise": "3.0.2",
    "es6-shim": "0.35.1",
    "leaflet": "0.7.7",
    "leaflet-label": "0.2.1-0",
    "lokijs": "1.3.16",
    "mapbox.js": "2.4.0",
    "milligram": "1.1.0",
    "normalize.css": "4.1.1",
    "reflect-metadata": "0.1.3",
    "requirejs": "2.2.0",
    "rxjs": "5.0.0-beta.0",
    "signalrjs": "0.2.2",
    "socket.io-client": "1.4.6",
    "system-json": "0.1.0",
    "systemjs": "0.19.17",
    "systemjs-plugin-json": "0.1.0",
    "zone.js": "0.5.11"
```

I was able to get the stack trace with more debugging of the boot startup of Angular 2:

```
"TypeError: decorator is not a function
    at DecoratePropertyWithoutDescriptor (http://localhost:3000/node_modules/angular2/bundles/angular2-polyfills.js:2916:13)
    at Object.decorate (http://localhost:3000/node_modules/angular2/bundles/angular2-polyfills.js:2443:20)
    at __decorate (http://localhost:3000/app/state/UserStore.js:6:96)
    at http://localhost:3000/app/state/UserStore.js:49:17
    at execute (http://localhost:3000/app/state/UserStore.js:54:14)
    at ensureEvaluated (http://localhost:3000/node_modules/systemjs/dist/system.src.js:3180:26)
    at ensureEvaluated (http://localhost:3000/node_modules/systemjs/dist/system.src.js:3172:11)
    at ensureEvaluated (http://localhost:3000/node_modules/systemjs/dist/system.src.js:3172:11)
    at ensureEvaluated (http://localhost:3000/node_modules/systemjs/dist/system.src.js:3172:11)
    at Object.execute (http://localhost:3000/node_modules/systemjs/dist/system.src.js:3301:13)"
```

**Comment by weichx** [comment]
I think the issue here is that you're using "module": "system". I havent worked with that at all but from looking at the docs it seems like adding allowSyntheticDefaultImports to tsConfig may allow interop between commonjs (which Cerialize expects) and system

**Comment by sondreb** [comment]
I tried to grab your .ts source, same problem.

And I tried to make my own simple and custom annotations, and they fail the same when being called with parameters, so the problem is likely not related to your great framework. Thanks for the help and I'll continue debugging!

PS: I will close the issue, but I'll come back and leave a comment on the fix when I find it, for future reference to others.

**Comment by weichx** [comment]
Sorry I couldn't help more! Good luck with the issue, I'll be interested to find out what it was

**Comment by sondreb** [comment]
The problem was that I had an attribute referencing a type that was declared below in the same source file. Doh.

See my last comment on the issue here: https://github.com/rbuckton/ReflectDecorators/issues/29

**Comment by cladera** [comment]
@sondreb glad you figure it out. However I'm afraid all of us will eventually came across this issue, and not always will have an easy solution. 
if you use some kind of generator, for instance, angular-cli, you'll barely have control over the declaration order. In my case, I hit an endpoint. This is what I want to accomplish:

```
// file /src/app/model/account.ts
export class Account {
  @autoserialize name: string;
  @autoserializeAs(Workspace) workspaces: Array<Workspace>;
}

// file /src/app/model/workspace.ts
export class Workspace {
  @autoserialize name: string;
  @autoserializeAs(Account) account: Account;
}
```

As you can see, even if I change the loading order, It'll crash as well.

**Comment by weichx** [comment]
This cannot work with annotations unless the library were written to handle some sort of async type loading. No matter what order you put things in either Workspace or Account will be undefined at the time the other class decorators are run, there is no way around that. The only way to make this work is to implement some other Api that can be run in a Serialize/Deserialize hook to re-gather the decorators. 

You still have the circular reference problem, which you cannot solve without a custom json schema that can support value lookups for other keys in the json file `{account: "Some Id defined elsewhere in your JSON`, which is out of scope for this project and probably not even a great idea. You're better off implementing an OnSerialize() method that will write an Id in the workspace's account field instead of a serialized account instance or vice versa. Your backend will have this same circular reference issue so you'd need to solve it there as well in a similar manner. 

TLDR: Circular references are going to be an issue for you even if it were possible to do what you are trying to accomplish. I'd recommend either 1.) Don't serialize workspaces as part of account, use ids instead. 2.) Don't serialize account as part of workspace, use ids instead.

**Comment by andregs** [comment]
Sometimes your instances don't have a circular dependency but your classes do. For example:

```typescript
// user.ts
export class User {
  @autoserialize name: string;
  @deserializeAs(News) newsFeed: News[];
}

// news.ts
export class News {
  @deserialize action: string;
  @deserializeAs(Date) when: Date;
  @deserializeAs(User) who: User;
}

// my-component.ts
john.newsFeed[0].who.name === 'Andrew'
```

It'd very useful if my _classes_ could reference each other this way. My instances don't depend on each other and I never serialize `newsFeed` back to DB, so there's no "circular dependency issue" on my app. However, the "decorator is not a function" error still happens because of `@deserializeAs(User) who: User`

**Comment by weichx** [comment]
I agree if this were possible it would be a great thing to have. Unfortunately, this is a JS problem and not a library one. The problem is that at parse time whichever one of your classes is parsed second, will not exist at the time the first one is parsed. C/C++ gets around this with forward declarations, C#/Java just have multi pass compilers that are smart enough to figure this out. Because JS is interpreted on the fly (ignore code generation for now), this is going to be a problem. 

I'll have to think about how to work around this

**Comment by jppellerin** [comment]
Another situation similar to this that I am running into is an object that holds some sort of inheritance of the same type.

```
export class MyType {
    @autoserialize id: string,
    @autoserializeAs(MyType) children: Array[MyType],
    @autoserialize otherAttribute: any
}
```

This will compile and the tests will run, but when loading into an Angular2 application, I get the same error.

This is where the error generates when trying to load the angular application (if this is of any help) : 
![image](https://cloud.githubusercontent.com/assets/4185657/20725735/d35dbdd2-b640-11e6-8e2d-9dc174834586.png)

**Comment by weichx** [comment]
I haven't touched angular 2 and am not familiar with how they do things but as far as the library goes your above code is correct. are you using commonjs or systemjs for modules?

**Comment by jppellerin** [comment]
We are targeting `commonjs`.

As a bit more information, this class is in a library that we export and then import in our Angular project.  Angular is then built using the angular-cli (therefore webpack).  It's there that it's failing when trying to load the application once built in a browser.  It seems to work fine when that library is loaded in a Node application.

**Comment by weichx** [comment]
Does everything work if you change the type of `children` to something other than `MyType`? I'm wondering if this is possibly an issue with using a type before its actually present in the runtime.

**Comment by jppellerin** [comment]
Everything works fine if I do this : 

```
export class MyType {
    @autoserialize id: string,
    @autoserialize children: Array<MyType>,
    @autoserialize otherAttribute: any
}
``` 

Except that the array items don't get deserialized - I basically get the `JSON.parse()` of the items.    (ie: If I was to add private variables with functions, etc - they would not exist).

**Comment by weichx** [comment]
right, and if you make children another type is it the same? Like some dummy class 
```
class Thing {
@autoserialize public value : string;
}

**Comment by bijeebuss** [comment]
Was there ever a workaround or fix discovered for this?

**Comment by weichx** [comment]
I'm not sure that there is much to be done as it's a language issue, not a code issue. You can try wrapping things in a function to be executed later

```typescript
class Thing {
@autoserializeAs(() => Stuff) public value : Stuff;
}
```
but I'm not sure if that helps
